src/decoder/stateful.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2023 The ChromiumOS Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+use thiserror::Error;
+use zerocopy::AsBytes;
+
+use crate::decoder::DecodedHandle;
+use crate::decoder::StreamInfo;
+use crate::DecodedFormat;
+use crate::Resolution;
+
+use std::sync::Arc;
+
+pub enum DecodeError {
+    InvalidState,
+    InvalidData,
+}
+
+pub enum ConfigError {
+    InvalidState,
+    NotSupported,
+    InvalidConfig,
+}
+
+/// Error returned by stateful backend methods.
+#[derive(Error, Debug)]
+pub enum StatefulBackendError {
+    #[error("not enough resources to proceed with the operation now")]
+    OutOfResources,
+    #[error("this format is not supported")]
+    UnsupportedFormat,
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+#[derive(Copy, Clone, Debug)]
```

Comment:
Since ccdec has a similar enum, we should probably have a single declaration in lib.rs since we already have `DecodedFormat` there.